File: backend/dotnet/nlayer-architecture/minimal/src/App.Template.Api/Controllers/NotificationsController.cs
using App.Template.Api.Models.Common;
using App.Template.Api.Repositories;
using App.Template.Api.Services;

using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace App.Template.Api.Controllers;

/// <summary>Notification management endpoints</summary>
[Authorize]
[ApiController]
[Route("api/notifications")]
public class NotificationsController : ControllerBase
{
    private readonly INotificationRepository _notificationRepository;
    private readonly ICurrentUserService _currentUserService;

    public NotificationsController(
        INotificationRepository notificationRepository,
        ICurrentUserService currentUserService)
    {
        _notificationRepository = notificationRepository;
        _currentUserService = currentUserService;
    }

    /// <summary>Get current user's notifications with pagination</summary>
    [HttpGet]
    public async Task<IActionResult> GetMyNotifications(
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 10,
        [FromQuery] bool? unreadOnly = null,
        [FromQuery] string? search = null,
        [FromQuery] string? sortBy = null,
        [FromQuery] string sortOrder = "desc",
        CancellationToken cancellationToken = default)
    {
        var userId = _currentUserService.UserId ?? "";

        var result = await _notificationRepository.GetByUserIdPagedAsync(
            userId, page, pageSize, unreadOnly, search, sortBy, sortOrder, cancellationToken);

        var dtoResult = new PagedResult<object>
        {
            Items = result.Items.Select(n => (object)new
            {
                n.Id,
                n.Title,
                n.Message,
                Type = n.Type.ToString().ToLower(),
                n.ReferenceId,
                n.ReferenceType,
                n.IsRead,
                n.CreatedAt
            }).ToList(),
            Pagination = result.Pagination
        };

        return Ok(new PaginatedResponse<object>
        {
            Success = true,
            Message = "Notifications retrieved successfully",
            Data = dtoResult.Items,
            Pagination = new PaginationInfo
            {
                Page = dtoResult.Pagination.Page,
                PageSize = dtoResult.Pagination.PageSize,
                TotalItems = dtoResult.Pagination.TotalItems,
                TotalPages = dtoResult.Pagination.TotalPages,
                HasNext = dtoResult.Pagination.HasNext,
                HasPrevious = dtoResult.Pagination.HasPrevious
            }
        });
    }

    /// <summary>Get unread notification count</summary>
    [HttpGet("unread-count")]
    public async Task<IActionResult> GetUnreadCount(CancellationToken cancellationToken)
    {
        var userId = _currentUserService.UserId ?? "";
        var count = await _notificationRepository.GetUnreadCountAsync(userId, cancellationToken);
        return Ok(ApiResponse.Ok(new { count }, "Unread count retrieved successfully"));
    }

    /// <summary>Mark a notification as read</summary>
    [HttpPut("{id:long}/read")]
    public async Task<IActionResult> MarkAsRead(long id, CancellationToken cancellationToken)
    {
        var userId = _currentUserService.UserId ?? "";
        var found = await _notificationRepository.MarkAsReadAsync(id, userId, cancellationToken);

        if (!found)
            return NotFound(ApiResponse.Fail($"Notification {id} not found"));

        return NoContent();
    }

    /// <summary>Mark all notifications as read</summary>
    [HttpPut("read-all")]
    public async Task<IActionResult> MarkAllAsRead(CancellationToken cancellationToken)
    {
        var userId = _currentUserService.UserId ?? "";
        await _notificationRepository.MarkAllAsReadAsync(userId, cancellationToken);
        return NoContent();
    }

    /// <summary>Delete a notification</summary>
    [HttpDelete("{id:long}")]
    public async Task<IActionResult> Delete(long id, CancellationToken cancellationToken)
    {
        var userId = _currentUserService.UserId ?? "";
        var deleted = await _notificationRepository.DeleteAsync(id, userId, cancellationToken);

        if (!deleted)
            return NotFound(ApiResponse.Fail($"Notification {id} not found"));

        return NoContent();
    }
}

File: backend/nestjs/nlayer-architecture/full/src/modules/notifications/notifications.controller.ts
import {
  Controller,
  Get,
  Put,
  Delete,
  Param,
  Query,
  ParseIntPipe,
  HttpCode,
  HttpStatus,
  UseGuards,
  Request,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth } from '@nestjs/swagger';
import { NotificationsService } from '../../services/notifications.service';
import { PaginationQueryDto } from '../../common/dto/pagination-query.dto';
import { ResponseMessage } from '../../common/decorators/response-message.decorator';

@ApiTags('Notifications')
@Controller('notifications')
@UseGuards(AuthGuard('jwt'))
@ApiBearerAuth('JWT-auth')
export class NotificationsController {
  constructor(private readonly notificationsService: NotificationsService) {}

  @Get()
  @ApiOperation({ summary: 'Get all notifications for current user' })
  @ApiResponse({ status: 200, description: 'Paginated list of notifications' })
  @ResponseMessage('Notifications retrieved successfully')
  async findAll(
    @Request() req: any,
    @Query() query: PaginationQueryDto,
    @Query('unreadOnly') unreadOnly?: string,
  ) {
    return this.notificationsService.findAllByUser(
      req.user.userId,
      query.page ?? 1,
      query.pageSize ?? 10,
      query.sortBy,
      query.sortOrder,
      unreadOnly === 'true',
    );
  }

  @Get('unread-count')
  @ApiOperation({ summary: 'Get unread notification count' })
  @ApiResponse({ status: 200, description: 'Unread count' })
  @ResponseMessage('Unread count retrieved')
  async getUnreadCount(@Request() req: any) {
    const count = await this.notificationsService.getUnreadCount(req.user.userId);
    return { count };
  }

  @Put(':id/read')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Mark notification as read' })
  @ApiResponse({ status: 204, description: 'Notification marked as read' })
  @ApiResponse({ status: 404, description: 'Notification not found' })
  async markAsRead(
    @Param('id', ParseIntPipe) id: number,
    @Request() req: any,
  ): Promise<void> {
    await this.notificationsService.markAsRead(id, req.user.userId);
  }

  @Put('read-all')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Mark all notifications as read' })
  @ApiResponse({ status: 204, description: 'All notifications marked as read' })
  async markAllAsRead(@Request() req: any): Promise<void> {
    await this.notificationsService.markAllAsRead(req.user.userId);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Delete a notification' })
  @ApiResponse({ status: 204, description: 'Notification deleted' })
  @ApiResponse({ status: 404, description: 'Notification not found' })
  async delete(
    @Param('id', ParseIntPipe) id: number,
    @Request() req: any,
  ): Promise<void> {
    await this.notificationsService.delete(id, req.user.userId);
  }
}
